enterprise/server/remote_execution/containers/firecracker/firecracker_test.go

```diff
@@ -137,7 +137,7 @@ func TestGuestAPIVersion(t *testing.T) {
 	// Note that if you go with option 1, ALL VM snapshots will be invalidated
 	// which will negatively affect customer experience. Be careful!
 	const (
-		expectedHash    = "a2ce5a10e715ce7233334ec2a06f3aaf4ae2238285f0d571677f9397a0ba62e1"
+		expectedHash    = "e35640804a9d339c6ae1cb9d0bd26f296a546e184b74e0514c4a724e480e86ff"
```

Comment:
any idea if this change is backwards-compatible with existing VM snapshots? will RPCs using the new proto succeed when they try to reach the old endpoint running within the VM?

(if it's not backwards compatible then we need to bump expectedVersion to 13)

Reply:
I will check, but I think the actual proto is byte-identical, which I believe should maintain interoperability?

Reply:
mmm, not identical but I think it just adds fields, which should still be fine, yes?

Reply:
okay, so, the `health.pb.go` file IS byte-identical, but `health_grpc.pb.go` is fairly different:

```diff
old mode 100755
new mode 100644
index 30a4a24f6..f96b8ab49
--- a/ours/health_grpc.pb.go
+++ b/theirs/health_grpc.pb.go
@@ -1,10 +1,27 @@
+// Copyright 2015 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The canonical version of this proto can be found at
+// https://github.com/grpc/grpc-proto/blob/master/grpc/health/v1/health.proto
+
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
 // versions:
-// - protoc-gen-go-grpc v1.3.0
-// - protoc             v5.26.0
-// source: proto/health.proto
+// - protoc-gen-go-grpc v1.5.1
+// - protoc             v5.27.1
+// source: grpc/health/v1/health.proto

-package health
+package grpc_health_v1

 import (
        context "context"
@@ -15,8 +32,8 @@ import (

 // This is a compile-time assertion to ensure that this generated file
 // is compatible with the grpc package it is being compiled against.
-// Requires gRPC-Go v1.32.0 or later.
-const _ = grpc.SupportPackageIsVersion7
+// Requires gRPC-Go v1.64.0 or later.
+const _ = grpc.SupportPackageIsVersion9

 const (
        Health_Check_FullMethodName = "/grpc.health.v1.Health/Check"
@@ -26,9 +43,37 @@ const (
 // HealthClient is the client API for Health service.
 //
 // For semantics around ctx use and closing/ending streaming RPCs, please refer to https://pkg.go.dev/google.golang.org/grpc/?tab=doc#ClientConn.NewStream.
+//
+// Health is gRPC's mechanism for checking whether a server is able to handle
+// RPCs. Its semantics are documented in
+// https://github.com/grpc/grpc/blob/master/doc/health-checking.md.
 type HealthClient interface {
+       // Check gets the health of the specified service. If the requested service
+       // is unknown, the call will fail with status NOT_FOUND. If the caller does
+       // not specify a service name, the server should respond with its overall
+       // health status.
+       //
+       // Clients should set a deadline when calling Check, and can declare the
+       // server unhealthy if they do not receive a timely response.
+       //
+       // Check implementations should be idempotent and side effect free.
        Check(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (*HealthCheckResponse, error)
-       Watch(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (Health_WatchClient, error)
+       // Performs a watch for the serving status of the requested service.
+       // The server will immediately send back a message indicating the current
+       // serving status.  It will then subsequently send a new message whenever
+       // the service's serving status changes.
+       //
+       // If the requested service is unknown when the call is received, the
+       // server will send a message setting the serving status to
+       // SERVICE_UNKNOWN but will *not* terminate the call.  If at some
+       // future point, the serving status of the service becomes known, the
+       // server will send a new message with the service's serving status.
+       //
+       // If the call terminates with status UNIMPLEMENTED, then clients
+       // should assume this method is not supported and should not retry the
+       // call.  If the call terminates with any other status (including OK),
+       // clients should retry the call with appropriate exponential backoff.
+       Watch(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (grpc.ServerStreamingClient[HealthCheckResponse], error)
 }

 type healthClient struct {
@@ -40,20 +85,22 @@ func NewHealthClient(cc grpc.ClientConnInterface) HealthClient {
 }

 func (c *healthClient) Check(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (*HealthCheckResponse, error) {
+       cOpts := append([]grpc.CallOption{grpc.StaticMethod()}, opts...)
        out := new(HealthCheckResponse)
-       err := c.cc.Invoke(ctx, Health_Check_FullMethodName, in, out, opts...)
+       err := c.cc.Invoke(ctx, Health_Check_FullMethodName, in, out, cOpts...)
        if err != nil {
                return nil, err
        }
        return out, nil
 }

-func (c *healthClient) Watch(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (Health_WatchClient, error) {
-       stream, err := c.cc.NewStream(ctx, &Health_ServiceDesc.Streams[0], Health_Watch_FullMethodName, opts...)
+func (c *healthClient) Watch(ctx context.Context, in *HealthCheckRequest, opts ...grpc.CallOption) (grpc.ServerStreamingClient[HealthCheckResponse], error) {
+       cOpts := append([]grpc.CallOption{grpc.StaticMethod()}, opts...)
+       stream, err := c.cc.NewStream(ctx, &Health_ServiceDesc.Streams[0], Health_Watch_FullMethodName, cOpts...)
        if err != nil {
                return nil, err
        }
-       x := &healthWatchClient{stream}
+       x := &grpc.GenericClientStream[HealthCheckRequest, HealthCheckResponse]{ClientStream: stream}
        if err := x.ClientStream.SendMsg(in); err != nil {
                return nil, err
        }
@@ -63,41 +110,59 @@ func (c *healthClient) Watch(ctx context.Context, in *HealthCheckRequest, opts .
        return x, nil
 }

-type Health_WatchClient interface {
-       Recv() (*HealthCheckResponse, error)
-       grpc.ClientStream
-}
-
-type healthWatchClient struct {
-       grpc.ClientStream
-}
-
-func (x *healthWatchClient) Recv() (*HealthCheckResponse, error) {
-       m := new(HealthCheckResponse)
-       if err := x.ClientStream.RecvMsg(m); err != nil {
-               return nil, err
-       }
-       return m, nil
-}
+// This type alias is provided for backwards compatibility with existing code that references the prior non-generic stream type by name.
+type Health_WatchClient = grpc.ServerStreamingClient[HealthCheckResponse]

 // HealthServer is the server API for Health service.
 // All implementations should embed UnimplementedHealthServer
-// for forward compatibility
+// for forward compatibility.
+//
+// Health is gRPC's mechanism for checking whether a server is able to handle
+// RPCs. Its semantics are documented in
+// https://github.com/grpc/grpc/blob/master/doc/health-checking.md.
 type HealthServer interface {
+       // Check gets the health of the specified service. If the requested service
+       // is unknown, the call will fail with status NOT_FOUND. If the caller does
+       // not specify a service name, the server should respond with its overall
+       // health status.
+       //
+       // Clients should set a deadline when calling Check, and can declare the
+       // server unhealthy if they do not receive a timely response.
+       //
+       // Check implementations should be idempotent and side effect free.
        Check(context.Context, *HealthCheckRequest) (*HealthCheckResponse, error)
-       Watch(*HealthCheckRequest, Health_WatchServer) error
-}
-
-// UnimplementedHealthServer should be embedded to have forward compatible implementations.
-type UnimplementedHealthServer struct {
-}
+       // Performs a watch for the serving status of the requested service.
+       // The server will immediately send back a message indicating the current
+       // serving status.  It will then subsequently send a new message whenever
+       // the service's serving status changes.
+       //
+       // If the requested service is unknown when the call is received, the
+       // server will send a message setting the serving status to
+       // SERVICE_UNKNOWN but will *not* terminate the call.  If at some
+       // future point, the serving status of the service becomes known, the
+       // server will send a new message with the service's serving status.
+       //
+       // If the call terminates with status UNIMPLEMENTED, then clients
+       // should assume this method is not supported and should not retry the
+       // call.  If the call terminates with any other status (including OK),
+       // clients should retry the call with appropriate exponential backoff.
+       Watch(*HealthCheckRequest, grpc.ServerStreamingServer[HealthCheckResponse]) error
+}
+
+// UnimplementedHealthServer should be embedded to have
+// forward compatible implementations.
+//
+// NOTE: this should be embedded by value instead of pointer to avoid a nil
+// pointer dereference when methods are called.
+type UnimplementedHealthServer struct{}

 func (UnimplementedHealthServer) Check(context.Context, *HealthCheckRequest) (*HealthCheckResponse, error) {
        return nil, status.Errorf(codes.Unimplemented, "method Check not implemented")
 }
-func (UnimplementedHealthServer) Watch(*HealthCheckRequest, Health_WatchServer) error {
+func (UnimplementedHealthServer) Watch(*HealthCheckRequest, grpc.ServerStreamingServer[HealthCheckResponse]) error {
        return status.Errorf(codes.Unimplemented, "method Watch not implemented")
 }
+func (UnimplementedHealthServer) testEmbeddedByValue() {}

 // UnsafeHealthServer may be embedded to opt out of forward compatibility for this service.
 // Use of this interface is not recommended, as added methods to HealthServer will
@@ -107,6 +172,13 @@ type UnsafeHealthServer interface {
 }

 func RegisterHealthServer(s grpc.ServiceRegistrar, srv HealthServer) {
+       // If the following call panics, it indicates UnimplementedHealthServer was
+       // embedded by pointer and is nil.  This will cause panics if an
+       // unimplemented method is ever invoked, so we test this at initialization
+       // time to prevent it from happening at runtime later due to I/O.
+       if t, ok := srv.(interface{ testEmbeddedByValue() }); ok {
+               t.testEmbeddedByValue()
+       }
        s.RegisterService(&Health_ServiceDesc, srv)
 }

@@ -133,21 +205,11 @@ func _Health_Watch_Handler(srv interface{}, stream grpc.ServerStream) error {
        if err := stream.RecvMsg(m); err != nil {
                return err
        }
-       return srv.(HealthServer).Watch(m, &healthWatchServer{stream})
-}
-
-type Health_WatchServer interface {
-       Send(*HealthCheckResponse) error
-       grpc.ServerStream
+       return srv.(HealthServer).Watch(m, &grpc.GenericServerStream[HealthCheckRequest, HealthCheckResponse]{ServerStream: stream})
 }

-type healthWatchServer struct {
-       grpc.ServerStream
-}
-
-func (x *healthWatchServer) Send(m *HealthCheckResponse) error {
-       return x.ServerStream.SendMsg(m)
-}
+// This type alias is provided for backwards compatibility with existing code that references the prior non-generic stream type by name.
+type Health_WatchServer = grpc.ServerStreamingServer[HealthCheckResponse]

 // Health_ServiceDesc is the grpc.ServiceDesc for Health service.
 // It's only intended for direct use with grpc.RegisterService,
@@ -168,5 +230,5 @@ var Health_ServiceDesc = grpc.ServiceDesc{
                        ServerStreams: true,
                },
        },
-       Metadata: "proto/health.proto",
+       Metadata: "grpc/health/v1/health.proto",
 }
```